crates/client/block_import/src/verify_apply.rs

```diff
@@ -293,3 +294,635 @@ fn block_hash(
+            block.unverified_global_state_root = None;
+            let validation = BlockValidationContext {
+                chain_id: ChainId::Mainnet,
+                ignore_block_order: false,
+                trust_global_tries: true,
+                trust_transaction_hashes: false,
+                trust_class_hashes: false,
+            };
+
+            let result = update_tries(&backend, &block, &validation, 1);
+            assert!(matches!(result.unwrap_err(), BlockImportError::Internal(_)));
+        }
+
+        /// Test error handling for mismatched global state root.
+        ///
+        /// Verifies that:
+        /// 1. The function returns a GlobalStateRoot error when the calculated root doesn't match the expected one.
+        #[tokio::test]
+        async fn test_update_tries_mismatch_global_state_root() {
+            let backend = setup_test_backend().await;
+            let mut block = create_dummy_block();
+            block.unverified_global_state_root = Some(felt!("0xb")); // Different from the calculated root
+            let validation = BlockValidationContext {
+                chain_id: ChainId::Mainnet,
+                ignore_block_order: false,
+                trust_global_tries: false,
+                trust_transaction_hashes: false,
+                trust_class_hashes: false,
+            };
+
+            let result = update_tries(&backend, &block, &validation, 1);
+            assert!(matches!(result.unwrap_err(), BlockImportError::GlobalStateRoot { .. }));
+        }
+
+        /// Test trie update with empty state diff.
+        ///
+        /// Verifies that:
+        /// 1. The function correctly handles an empty state diff.
+        /// 2. It returns the expected global state root (0x0 in this case).
+        #[tokio::test]
+        async fn test_update_tries_empty_state_diff() {
+            let backend = setup_test_backend().await;
+            let mut block = create_dummy_block(); // StateDiff is already empty by default
+            block.unverified_global_state_root = Some(felt!("0x0"));
+            let validation = BlockValidationContext {
+                chain_id: ChainId::Mainnet,
+                ignore_block_order: false,
+                trust_global_tries: false,
+                trust_transaction_hashes: false,
+                trust_class_hashes: false,
+            };
+
+            let result = update_tries(&backend, &block, &validation, 1).expect("Update tries should succeed");
+            assert_eq!(result, felt!("0x0"), "Update tries should succeed with empty state diff");
+        }
+    }
+
+    mod block_hash_tests {
+        use super::*;
+        /// Test successful block hash calculation.
+        ///
+        /// Verifies that:
+        /// 1. The function correctly calculates the block hash.
+        /// 2. The returned header contains the expected block number.
+        #[test]
+        fn test_block_hash_success() {
+            let block = PreValidatedBlock {
```

Comment:
can you `create_dummy_block` by moving it outside the module?

Reply:
resolved